### ISSUE TITLE
[java] ReturnEmptyCollectionRatherThanNull: False negatives when a returned expression can evaluate to null

### ISSUE THREAD

**Issue by Chordrain** [pmd/pmd#6744]
**PMD Version:** 7.24.0
**Category:** [ReturnEmptyCollectionRatherThanNull](https://pmd.github.io/pmd/pmd_rules_java_errorprone.html#returnemptycollectionratherthannull)
**Severity:** Warning (Error Prone)

## Summary

PMD's `ReturnEmptyCollectionRatherThanNull` rule flags `return null;` statements inside methods whose declared return type is a `Collection` (or one of its subtypes) or an array. The matcher only recognizes a bare `return <NullLiteral>;` and does not look through a constant-condition ternary.

Wrapping the null in a constant-true ternary — `return (true ? null : null);` — silences the rule (false negative), even though the method still unconditionally returns the null literal on every execution path (the ternary's value is `null` per JLS §15.25).

### before — bare `return null;` (1 finding)

```java
import java.util.List;

public class ReturnNullBefore {
    public List<String> getList() {
        return null;                       // flagged
    }
}
```

### after — ternary-wrapped null (0 findings)

```java
import java.util.List;

public class ReturnNullAfter {
    public List<String> getList() {
        return (true ? null : null);       // FALSE NEGATIVE: not flagged
    }
}
```

## Expected behavior

`ReturnEmptyCollectionRatherThanNull` should report the same violation on BEFORE and AFTER. Both methods are declared to return `List<String>` and both return the null literal on every path; the constant-true ternary `(true ? null : null)` is a literal-null expression per JLS §15.25.

## Actual behavior

- `return null;` → 1 finding: `Return an empty collection rather than null.` (correct).
- `return (true ? null : null);` → 0 findings, even though the method still unconditionally returns `null`.

The matcher inspects only a bare `return <NullLiteral>;` AST and misses null returns wrapped in a parenthesised / constant-condition ternary.

**Comment by adangel** [comment]
> the constant-true ternary `(true ? null : null)`

This looks contrived. Did you see actual code that uses this construct? (It would qualify in a rule on its own - unnecessarily complicated way of writing null)

I'm not sure, if the rule should concentrate on this very specific, special case.

What about:

```
return someCondition() ? aList : null;
```

Is this already covered? I don't think so.

The rule should probably be improved to check, whether null would be returned via a ternary operator (regardless of the condition).

**Comment by Chordrain** [comment]
Fair point. My intention is not to argue that `(true ? null : null)` is a realistic pattern by itself. It was simply the smallest reproducer I found.

The broader issue seems to be that the rule only detects a bare `return null;` and does not inspect conditional expressions that may return `null`. For example:

```java
import java.util.List;

public class Sample {
    public List<String> getList() {
        List<String> aList;
        return false ? aList : null;
    }
}
```

This also produces no warning.

So the problem appears to be wider than the constant-true ternary case: null-returning branches within conditional expressions are currently missed by the rule.

**Comment by adangel** [comment]
> It was simply the smallest reproducer I found.

Sure. The concern I have is, that this might be misleading, if someone tries to fix this - one might concentrate only on this simple case and miss the underlying problem.

> So the problem appears to be wider than the constant-true ternary case: null-returning branches within conditional expressions are currently missed by the rule.

exactly
